crypto/s2n_ecdsa.c

```diff
@@ -71,123 +78,91 @@ int s2n_ecdsa_verify(const struct s2n_ecdsa_public_key *key, struct s2n_hash_sta
-    uint32_t parsed_len = cert_to_parse - asn1der->data;
+    uint8_t input[16];
+    struct s2n_blob random_input;
+    struct s2n_blob signature;
+    struct s2n_hash_state state_in, state_out;
 
-    if (parsed_len != asn1der->size) {
-        X509_free(cert);
-        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
-    }
+    random_input.data = input;
+    random_input.size = sizeof(input);
+    GUARD(s2n_get_public_random_data(&random_input));
 
-    EVP_PKEY *public_key = X509_get_pubkey(cert);
-    X509_free(cert);
+    /* s2n_hash_new only allocates memory when using high-level EVP hashes, currently restricted to FIPS mode. */
+    GUARD(s2n_hash_new(&state_in));
+    GUARD(s2n_hash_new(&state_out));
 
-    if (public_key == NULL) {
-        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
-    }
+    GUARD(s2n_hash_init(&state_in, S2N_HASH_SHA1));
+    GUARD(s2n_hash_init(&state_out, S2N_HASH_SHA1));
+    GUARD(s2n_hash_update(&state_in, input, sizeof(input)));
+    GUARD(s2n_hash_update(&state_out, input, sizeof(input)));
 
-    if (EVP_PKEY_base_id(public_key) != EVP_PKEY_EC) {
-        EVP_PKEY_free(public_key);
-        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
-    }
+    const s2n_ecdsa_private_key *priv_key = &priv->key.ecdsa_key;
+    GUARD(s2n_alloc(&signature, s2n_ecdsa_signature_size(priv_key)));
+    
+    GUARD(s2n_ecdsa_sign(priv, &state_in, &signature));
+    GUARD(s2n_ecdsa_verify(pub, &state_out, &signature));
 
-    key->eckey = EVP_PKEY_get1_EC_KEY(public_key);
-    if (key->eckey == NULL) {
-        EVP_PKEY_free(public_key);
-        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
-    }
+    GUARD(s2n_hash_free(&state_in));
+    GUARD(s2n_hash_free(&state_out));
 
-    EVP_PKEY_free(public_key);
+    return 0;
+}
 
+static int s2n_ecdsa_key_free(struct s2n_pkey *pkey)
+{
+    struct s2n_ecdsa_key *ecdsa_key = &pkey->key.ecdsa_key;
+    notnull_check(ecdsa_key->ec_key);
+    EC_KEY_free(ecdsa_key->ec_key);
     return 0;
 }
 
-int s2n_asn1der_to_ecdsa_private_key(struct s2n_ecdsa_private_key *key, struct s2n_blob *asn1der)
+int s2n_ecdsa_signature_size(const s2n_ecdsa_private_key *key)
 {
-    uint8_t *key_to_parse = asn1der->data;
+    notnull_check(key->ec_key);
 
-    EVP_PKEY *pkey = d2i_PrivateKey(EVP_PKEY_EC, NULL, (const unsigned char **)(void *)&key_to_parse, asn1der->size);
-    if (pkey == NULL) {
-        S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
-    }
-    
-    EC_KEY *ec_key = EVP_PKEY_get1_EC_KEY(pkey);
-    EVP_PKEY_free(pkey);
-    if (ec_key == NULL) {
-        S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
-    }
+    return ECDSA_size(key->ec_key);
+}
 
-    /* If key parsing is successful, d2i_PrivateKey increments *key_to_parse to the byte following the parsed data */
-    uint32_t parsed_len = key_to_parse - asn1der->data;
-    if (parsed_len != asn1der->size) {
+int s2n_evp_pkey_to_ecdsa_private_key(s2n_ecdsa_private_key *ecdsa_key, EVP_PKEY *evp_private_key)
```

Comment:
There is a subtle difference in the error messages used in each function. I could converge them and use a more generic "Error decoding key" message fairly easily. I made that change, and went to go fix the equivalent RSA functions, but realized that the RSA functions can't converge because there are also minor functional differences for public and private keys. (RSA_check_key can't be called on public keys). My personal preference would be to leave them as separate functions for now to keep the RSA and ECDSA cases "symmetrical". If you have a strong opinion to the contrary, let me know.